Mathlib/Tactic/NoncommRing.lean

```diff
@@ -53,8 +60,13 @@ macro_rules
           -- Pull `zsmul n` out the front so `abel` can see them.
           mul_smul_comm, smul_mul_assoc,
           -- Pull out negations.
-          neg_mul, mul_neg] |
+          neg_mul, mul_neg,
+          -- user-specified simp lemmas
+          $rules',*] |
         fail "`noncomm_ring` simp lemmas don't apply; try `abel` instead") <;>
-      first | abel1 | abel_nf))
+      first | abel1 | abel_nf)
+    -- if a manual rewrite rule is provided, we repeat the tactic
+    -- (since abel might simplify and allow the rewrite to apply again)
+    if rules.isSome then `(tactic| repeat1 ($tac;)) else `(tactic| $tac)
```

Comment:
I'm ever so slightly concerned that this will make things slow because we'll be calling `abel` multiple times, but perhaps this concern is overblown because in practice it won't happen enough times to matter.

Reply:
I agree that this is not the most efficient implementation. For this reason I made sure that `noncomm_ring` without arguments will not use `repeat`. I think that `noncomm_ring` with arguments will be exceedingly rare that it should not be an issue.